FF145 Storage Access HTTP headers (#41614)
mdn/content: files/en-us/web/http/reference/headers/activate-storage-access/index.md

---
title: Activate-Storage-Access header
short-title: Activate-Storage-Access
slug: Web/HTTP/Reference/Headers/Activate-Storage-Access
page-type: http-header
browser-compat: http.headers.Activate-Storage-Access
sidebar: http
---

The HTTP **`Activate-Storage-Access`** {{glossary("response header")}} allows a server to activate a granted permission to access its [unpartitioned cookies](/docs/Web/API/Storage_Access_API#unpartitioned_versus_partitioned_cookies) in a cross-site request.

The server relies on permission state information provided in the request's {{httpheader("Sec-Fetch-Storage-Access")}} header.

The headers are collectively referred to as [storage access headers](/en-US/docs/Web/API/Storage_Access_API#storage_access_headers).

> They provide an efficient alternative to first loading the resource without cookies and using the [Storage Access API](/en-US/docs/Web/API/Storage_Access_API#) to activate the permission.

Reviewer: ```suggestion
They provide an efficient alternative to first loading the resource without cookies, using the [Storage Access API](/en-US/docs/Web/API/Storage_Access_API#) to activate the permission, and then reloading the resource with cookies.
```

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>{{Glossary("Fetch Metadata Request Header")}}</td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden request header")}}</th>
      <td>No</td>
    </tr>
    <tr>
      <th scope="row">
        {{Glossary("CORS-safelisted request header")}}
      </th>
      <td>No</td>
    </tr>
  </tbody>
</table>

## Syntax

```http
Activate-Storage-Access: retry; allowed-origin="https://foo.bar"
Activate-Storage-Access: retry; allowed-origin=*
Activate-Storage-Access: load
```

## Directives

- `retry`

>   - : Request the browser activate the storage-access permission for the context and then retry the request with cookies included.

Reviewer: ```suggestion
  - : Request that the browser activate the storage-access permission for the context and then retry the request with cookies included.
```

>     The `allowed-origin` parameter must be specified to whitelist the specific origin (or specify `*` to allow any origin).

Reviewer: ```suggestion
    The `allowed-origin` parameter must be specified to allow the specific origin (specify `*` to allow any origin).
```

We don't use "whitelist" anymore.

- `load`

>   - : Request the user agent activate the store-access permission for the context and then load the resource.

Reviewer: ```suggestion
  - : Request that the browser activate the store-access permission for the context and then load the resource.
```

The resource must first be requested without cookies, so the server will return an uncredentialed version of the resource that will not have access to its own cookies.
After loading, this resource can call {{domxref("Document.requestStorageAccess()")}} with transient activation to request the storage-access permission.
If granted by the user, the permission is stored by the browser to a key associated with the embedder and embedded _site_.
The resource must then reload the resource, which it is now able to request with cookies because it has the `active` permission state for the current context.

The permission is granted for a particular embedder/embedded site, but only activated for a particular context, such as a frame, or browser tab.
What this means is that if you load the same page in a new tab or frame, for example, the permission state of that context will be `inactive`; it won't become `active` until the permission is activated.

> The normal storage access flow is to again request the resource without cookies, call `Document.requestStorageAccess()` to activate the existing permission, then reload the resource with cookies.

Reviewer: This all feels a bit repetitive. Could you link to the same info on the main Storage Access API page to avoid repeating it here?

Reviewer: I made some grammatical tweaks to this.

A server can use `Activate-Storage-Access` to activate a granted permission for a context, avoiding the need to load the resource just so that it can activate the permission by calling `Document.requestStorageAccess()`.

The way this works is that the browser adds `Sec-Fetch-Storage-Access` to requests to indicate the storage access state.
If it adds `Sec-Fetch-Storage-Access: inactive` to requests, along with the `Origin` header indicating the source of the request, the server knows that the resource has permission but that it isn't active.
The server can then respond with `Activate-Storage-Access: retry; allowed-origin="<request_origin>"` to ask the browser to activate the permission for the context and retry the request.

> The browser then sends the request again, this time with `Sec-Fetch-Storage-Access: active` and including cookies, and the server responds with the credentialed version of the resource, which has access to its cookies as though it was a first party resource.

Reviewer: ```suggestion
The browser then sends the request again, this time with `Sec-Fetch-Storage-Access: active` and including cookies, and the server responds with the credentialed version of the resource, which has access to its cookies as though it were a first-party resource.
```

> Note that the response must also the {{httpheader("Vary")}} header with `Sec-Fetch-Storage-Access`.

Reviewer: ```suggestion
The response must also include the {{httpheader("Vary")}} header with `Sec-Fetch-Storage-Access`.
```

You don't mention this at all on the Storage Access API main page. I think maybe you should.

Author: I will add!

They use the context of a site `https://mysite.example` that includes an {{htmlelement("iframe")}} that embeds the user profile page `embedded.com/user/profile`.

### Request when secure-access permission initially not granted

The request below assumes that this is the first time that the user has visited a page that embeds anything from `embedded.com`, so the storage access permission has not been granted.

The request is for a cross-site iframe made with credentials mode ["include"](/en-US/docs/Web/API/Request/credentials#include).
The browser has added `Sec-Fetch-Storage-Access: none` to the request, because the `secure-access` permission has not been granted.
It hasn't added cookies because they blocked by default.

```http
GET /user/profile HTTP/1.1
Host: embedded.com
Origin: https://mysite.example
Sec-Fetch-Dest: iframe
Sec-Fetch-Site: cross-site
Sec-Fetch-Mode: navigate
Sec-Fetch-Storage-Access: none
Credentials-Mode: include
```

The server can't ask the browser to activate a granted permission using `Activate-Storage-Access`, because no permission has been granted.

> Instead the server returns a non-credentialed version of the resource.

Reviewer: ```suggestion
Instead, the server returns a non-credentialed version of the resource.
```

The server does add the {{httpheader("Vary")}} header, as the response may change with `Sec-Fetch-Storage-Access`.

```http
HTTP/1.1 200 OK
Content-Type: text/html
Vary: Sec-Fetch-Storage-Access

<html>
  ...
</html>
```

> The embedded page would call {{domxref("Document.requestStorageAccess()")}} with transient activation to request the storage-access permission (see the [Storage Access API](/en-US/docs/Web/API/Storage_Access_API) for more information).

Reviewer: Nice work. This makes more sense now.

> If granted the embedded page would have both a granted and active storage-access permission.

Reviewer: ```suggestion
If permission is granted, the embedded page would have both a granted and active storage-access permission.
```

It would then reload itself, resulting in the following request.

> This time the browser adds `Sec-Fetch-Storage-Access: active` and includes the third party cookies, reflecting the permission state of the embedded content.

Reviewer: ```suggestion
This time the browser adds `Sec-Fetch-Storage-Access: active` and includes the third-party cookies, reflecting the permission state of the embedded content.
```

```http
GET /user/profile HTTP/1.1
Host: embedded.com
Origin: https://mysite.example
Sec-Fetch-Dest: iframe
Sec-Fetch-Site: cross-site
Sec-Fetch-Mode: navigate
Sec-Fetch-Storage-Access: active
Credentials-Mode: include
Cookie: sessionid=abc123
```

The server responds with the credentialed version of the resource, which may be different than the first version that was loaded, and adds the header `Activate-Storage-Access: load`.
The browser loads the page, which will now have access to its own cookie information.

```http
HTTP/1.1 200 OK
Content-Type: text/html
Vary: Sec-Fetch-Storage-Access
Activate-Storage-Access: load

<html>
  ...
</html>
```

Note that when working with a server that does not use these headers, the flow would be the same.
This is because a resource always has to be loaded the first time without cookies before it can be granted permission.

### Server activating a permission

> This example uses the same scenario as in the previous example, but now we assume that the context has permission but that it has not yet been activated (with the API we'd activate the permission by reloading the resource so it can call `Document.requestStorageAccess()`).

Reviewer: ```suggestion
This example uses the same scenario as in the previous example, but now we assume that the context has permission, but it has not yet been activated. (With the API, we'd activate the permission by reloading the resource so it can call `Document.requestStorageAccess()`.)
```
